validator/client/duplicate_detection.go

```diff
@@ -0,0 +1,248 @@
+	if err != nil {
+		return err
+	}
+
+	// Get the currentEpoch and genesisEpoch
+	slot := <-v.NextSlot()
+	currentEpoch := helpers.SlotToEpoch(slot)
+	genesisEpoch := params.BeaconConfig().GenesisEpoch
+
+	// Counting N epochs from the starting Slot(substract 1 since we alwasy check slot-1 at the start).
+	endingSlot := slot.Add(uint64(params.BeaconConfig().SlotsPerEpoch.Mul(uint64(NoEpochsToCheck)))).Sub(1)
+
+	// Ensure currentEpoch > genesisEpoch
+	// If the current is equal to the genesis or prior to genesis then no duplicate check is performed.
+	if genesisEpoch >= currentEpoch {
+		return nil
+	}
+
+	// Either a proposal or attestation duplicate is detected at one of the slots in a 2 epoch period which results
+	// in a forced validator stop, or none is found and flow continues in the validator runner.
+	// Steps:
+	// 1. Detect doppelganger.
+	// 2. if not found sleep till next slot; Go to 4
+	// 3. If found exit
+	// 4. repeat for 2 epochs, Go to 1.
+	for {
+		// Are we done?
+		if slot >= endingSlot.Sub(1) {
+			log.Info("Doppelganger service - finished the epoch checks for duplicates ")
+			// not found. Return nil struct to signal for the Duplicate Check goroutine in node to shut down only.
+			ret := &DuplicateDetection{}
+			ret.DuplicateKey = nil
+			v.duplicateFeed.Send(ret)
+			return nil
+		}
+
+		log.Infof("Doppelganger service - starts the check at previous slot %d", slot-1)
+		// Detect a doppelganger in the previous Slot
+		dupIndex, err := v.detectDoppelganger(ctx, slot-1)
+		if err != nil {
+			return err
+		}
+		if dupIndex != nil {
+			log.Infof("Doppelganger detected! Validator key 0x%x seems to be running elsewhere."+
+				"This process will exit, avoiding a proposer or attester slashing event."+
+				"Please ensure you are not running your validator in two places simultaneously.", dupIndex)
+
+			// Broadcast the findings at the slot. The Node process will listen and shutdown.
+			ret := &DuplicateDetection{}
+			ret.DuplicateKey = dupIndex
+			ret.Slot = slot
+			v.duplicateFeed.Send(ret)
+			return nil
+		}
+
+		// Sleep time between now and start of next slot.
+		nextSlotTime := v.SlotDeadline(slot)
+		timeRemaining := time.Until(nextSlotTime)
+		// Still time till next slot? sleep through and loop again
+		if timeRemaining >= 0 {
+			log.WithFields(logrus.Fields{
+				"timeRemaining": timeRemaining,
+			}).Info("Sleeping until the next slot - Doppelganger service")
+			time.Sleep(timeRemaining)
+			continue
+		} // else condition should not happen.Clock is off?
+		// Or maybe Detection process at previous slot took too long
+
+		// Get next slot
+		slot = <-v.NextSlot()
+		log.Infof("Doppelganger service - new Slot %d", slot)
+	}
+
+}
+
+// Doppelganger detection
+// Retrieve all the beacon-chain blocks at the given slot
+// and check for blocks with the same pubKey as this validator
+func (v *validator) detectDoppelganger(ctx context.Context, slot types.Slot) ([]byte, error) {
```

Comment:
This seems like a very inefficient approach here. We fetch data from the beacon node, have it marshalled/unmarshalled over the wire and then check whether it was slashable validator here. The beacon node is much better equipped to handle these sort of checks, which is why I mentioned having an rpc method in the beacon node for this. An example:

https://github.com/prysmaticlabs/prysm/blob/develop/beacon-chain/rpc/validator/status.go#L32

You would need to specify a new v1alpha1 rpc endpoint here, and then use that for validator communication. 
https://github.com/prysmaticlabs/ethereumapis/blob/master/eth/v1alpha1/validator.proto

You would need to decide on the rpc request structure, etc. 

The actual process of checking the slashability of any validator can be carried out by that particular rpc endpoint. So the endpoint becomes a blackbox through which you can employ methods to check validator whether a validator has attested over N epochs. 

You can keep a lot of this logic out of the validator, the validator for most purposes should be as light-weight as possible.

Reply:
Thanks, needed a review!  i knew i was doing it in a safe but absurd brute way.